google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddDynamicPageFeed.java

```diff
@@ -0,0 +1,540 @@
+    try (FeedItemServiceClient feedItemServiceClient =
+        googleAdsClient.getLatestVersion().createFeedItemServiceClient()) {
+      // Adds the feed items.
+      MutateFeedItemsResponse response =
+          feedItemServiceClient.mutateFeedItems(Long.toString(customerId), operations);
+      // Displays the results.
+      for (MutateFeedItemResult result : response.getResultsList()) {
+        System.out.printf(
+            "Created feed items with resource name '%s'.%n", result.getResourceName());
+      }
+    }
+    return;
+  }
+
+  /**
+   * Updates a campaign to set the DSA feed.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID in which to create criterion.
+   * @param feedDetails the relevant information about the feed and its attributes.
+   * @param campaignId the campaign ID of the campaign to update.
+   */
+  private void updateCampaignDsaSetting(
+      GoogleAdsClient googleAdsClient, long customerId, FeedDetails feedDetails, long campaignId) {
+    DynamicSearchAdsSetting dsaSetting =
+        DynamicSearchAdsSetting.newBuilder(getDsaSetting(googleAdsClient, customerId, campaignId))
+            .addFeeds(StringValue.of(feedDetails.getFeedResourceName()))
+            .build();
+
+    // Creates the campaign.
+    Campaign campaign =
+        Campaign.newBuilder()
+            .setResourceName(ResourceNames.campaign(customerId, campaignId))
+            .setDynamicSearchAdsSetting(dsaSetting)
+            .build();
+    // Creates the operation.
+    CampaignOperation operation =
+        CampaignOperation.newBuilder()
+            .setUpdate(campaign)
+            .setUpdateMask(FieldMasks.allSetFieldsOf(campaign))
+            .build();
+    // Creates the campaign service.
+    try (CampaignServiceClient campaignServiceClient =
+        googleAdsClient.getLatestVersion().createCampaignServiceClient()) {
+      // Updates the campaign.
+      MutateCampaignsResponse response =
+          campaignServiceClient.mutateCampaigns(
+              Long.toString(customerId), ImmutableList.of(operation));
+      // Displays the results.
+      for (MutateCampaignResult mutateCampaignResult : response.getResultsList()) {
+        System.out.printf(
+            "Updated campaign with resourceName: %s.%n", mutateCampaignResult.getResourceName());
+      }
+    }
+    return;
+  }
+
+  /**
+   * Returns the DSA settings for a campaign. Throws an error if the campaign does not exist or is
+   * not a DSA campaign.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID in which to create criterion.
+   * @param campaignId the campaign ID.
+   */
+  private DynamicSearchAdsSetting getDsaSetting(
+      GoogleAdsClient googleAdsClient, long customerId, long campaignId) {
+    // Creates the query.
+    // You must request all DSA fields in order to update the DSA settings in the following step.
+    String query =
```

Comment:
Why do we need this round trip? Can't we just overwrite the existing DSA setting?

Reply:
Yes, you can overwrite the existing DSA settings. The potential downside is that you have to know the values for domain_name, language_code, and use_supplied_urls_only you used when originally creating the campaign if you want them to remain the same. Presumably you would want the values to remain the same because you set those values initially. It seems counter-intuitive to me to have to look those up or introduce the risk of potentially making an inadvertent change to the DB. In my opinion that inconvenience/risk outweighs the cost of the additional roundtrip, so I'll leave it as for now. That said, it's an easy fix, so LMK if you want to disagree, and I can change it.